### PR TITLE
Add neighbors and path GraphQL queries.

### PR DESCRIPTION
**Pull request by mihaimaruseac**
These allow extract node neighborhood information, where the node type is not as relevant as the connections between them.

Renamed the ID interface, added a neighbors field, implemented it for all currently defined structs.

Added the 2 GraphQL queries and resolvers.

```gql
query neighbors {
  neighbors(node: 47) {
    __typename
    ... on Package {
      ...allPkgTree
    }
    ... on IsDependency {
      id
    }
  }
}

query path {
  path(subject:9, target:43, maxPathLength:30) {
    __typename
    ... on Package {
      ...allPkgTree
    }
    ... on IsDependency {
      id
    }
  }
}
```

```gql
{
  "data": {
    "path": [
      {
        "__typename": "Package",
        "id": "42",
        "type": "deb",
        "namespaces": [
          {
            "id": "43",
            "namespace": "ubuntu",
            "names": []
          }
        ]
      },
      {
        "__typename": "Package",
        "id": "42",
        "type": "deb",
        "namespaces": [
          {
            "id": "43",
            "namespace": "ubuntu",
            "names": [
              {
                "id": "44",
                "name": "dpkg",
                "versions": []
              }
            ]
          }
        ]
      },
      {
        "__typename": "Package",
        "id": "42",
        "type": "deb",
        "namespaces": [
          {
            "id": "43",
            "namespace": "ubuntu",
            "names": [
              {
                "id": "44",
                "name": "dpkg",
                "versions": [
                  {
                    "id": "45",
                    "version": "1.19.0.4",
                    "qualifiers": [
                      {
                        "key": "arch",
                        "value": "amd64"
                      }
                    ],
                    "subpath": ""
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        "__typename": "IsDependency",
        "id": "47"
      },
      {
        "__typename": "Package",
        "id": "9",
        "type": "conan",
        "namespaces": [
          {
            "id": "10",
            "namespace": "openssl.org",
            "names": [
              {
                "id": "11",
                "name": "openssl",
                "versions": []
              }
            ]
          }
        ]
      },
      {
        "__typename": "Package",
        "id": "9",
        "type": "conan",
        "namespaces": [
          {
            "id": "10",
            "namespace": "openssl.org",
            "names": []
          }
        ]
      },
      {
        "__typename": "Package",
        "id": "9",
        "type": "conan",
        "namespaces": []
      }
    ]
  }
}
```

and

```gql
{
  "data": {
    "neighbors": [
      {
        "__typename": "Package",
        "id": "42",
        "type": "deb",
        "namespaces": [
          {
            "id": "43",
            "namespace": "ubuntu",
            "names": [
              {
                "id": "44",
                "name": "dpkg",
                "versions": [
                  {
                    "id": "45",
                    "version": "1.19.0.4",
                    "qualifiers": [
                      {
                        "key": "arch",
                        "value": "amd64"
                      }
                    ],
                    "subpath": ""
                  }
                ]
              }
            ]
          }
        ]
      },
      {
        "__typename": "Package",
        "id": "9",
        "type": "conan",
        "namespaces": [
          {
            "id": "10",
            "namespace": "openssl.org",
            "names": [
              {
                "id": "11",
                "name": "openssl",
                "versions": []
              }
            ]
          }
        ]
      }
    ]
  }
}
```

### CODE REVIEW

File: pkg/assembler/backends/testing/cve.go [hunk 1]
```diff
@@ -70,6 +70,28 @@ type cveIDNode struct {
...
+func (n *cveNode) neighbors() []uint32 {
+	out := make([]uint32, 0, len(n.cveIDs))
+	for _, v := range n.cveIDs {
+		out = append(out, v.id)
+	}
+	return out
+}
+
+func (n *cveIDNode) neighbors() []uint32 {
+	// TODO: DO NOT MERGE: add backedge
```
Comment (jeffmendoza) [inline pkg/assembler/backends/testing/cve.go:82]: > DO NOT MERGE

oops =)

Comment (mihaimaruseac) [reply pkg/assembler/backends/testing/cve.go:83]: Yeah, I have it to know to integrate the new PRs after rebasing :)

File: pkg/assembler/backends/testing/isOccurrence.go [hunk 2]
```diff
@@ -44,6 +44,25 @@ type isOccurrenceStruct struct {
...
+func (n *isOccurrenceStruct) neighbors() []uint32 {
+	out := make([]uint32, 0, 3)
+	// TODO: replace maxUint32 with 0
+	if n.pkg != maxUint32 {
+		out = append(out, n.pkg)
+	}
+	if n.source != maxUint32 {
+		out = append(out, n.source)
+	}
+	if n.artifact != maxUint32 {
```
Comment (jeffmendoza) [inline pkg/assembler/backends/testing/isOccurrence.go:56]: Artifact should always be set

File: pkg/assembler/backends/testing/artifact.go [hunk 3]
```diff
@@ -40,6 +40,17 @@ type artStruct struct {
 
 func (n *artStruct) getID() uint32 { return n.id }
 
+func (n *artStruct) neighbors() []uint32 {
+	out := make([]uint32, 0, len(n.hashEquals)+len(n.occurrences))
+	out = append(out, n.hashEquals...)
+	out = append(out, n.occurrences...)
```
Comment (jeffmendoza) [inline pkg/assembler/backends/testing/artifact.go:46]: I snuk in hasSLSAs before you.

File: pkg/assembler/backends/testing/isOccurrence.go (hunk 2 above)
Comment (mihaimaruseac) [reply pkg/assembler/backends/testing/isOccurrence.go:56]: Thanks, updated

**Review by jeffmendoza — APPROVED** [review]

File: pkg/assembler/backends/testing/path.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+	nodeMap := map[uint32]dfsNode{}
+
+	nodeMap[from] = dfsNode{}
+	queue = append(queue, from)
+
+	found := false
+	for len(queue) > 0 {
+		now := queue[0]
+		queue = queue[1:]
```
Comment (pxp928) [inline pkg/assembler/backends/testing/path.go:98]: nice!

**Review by pxp928 — APPROVED** [review]
LGTM! Just a question

File: pkg/assembler/backends/backends.go [hunk 5]
```diff
@@ -71,6 +71,10 @@ type Backend interface {
 	IngestHasSourceAt(ctx context.Context, pkg model.PkgInputSpec, pkgMatchType model.MatchFlags, source model.SourceInputSpec, hasSourceAt model.HasSourceAtInputSpec) (*model.HasSourceAt, error)
 	IngestIsVulnerability(ctx context.Context, osv model.OSVInputSpec, vulnerability model.CveOrGhsaInput, isVulnerability model.IsVulnerabilityInputSpec) (*model.IsVulnerability, error)
 	IngestVEXStatement(ctx context.Context, subject model.PackageOrArtifactInput, vulnerability model.CveOrGhsaInput, vexStatement model.VexStatementInputSpec) (*model.CertifyVEXStatement, error)
+
+	// More advanced queries
+	Path(ctx context.Context, subject string, target string, maxPathLength int) ([]model.Node, error)
+	Neighbors(ctx context.Context, node string) ([]model.Node, error)
```
Comment (lumjjb) [inline pkg/assembler/backends/backends.go:77]: nit should this be uint32?

**Review by lumjjb — APPROVED** [review]

File: pkg/assembler/backends/testing/path.go [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+	for now != from {
+		reversedPath = append(reversedPath, now)
+		now = nodeMap[now].parent
+	}
+	reversedPath = append(reversedPath, now)
+
+	// reverse path
+	path := make([]uint32, len(reversedPath))
+	for i, x := range reversedPath {
+		path[len(reversedPath)-i-1] = x
```
[the remark this answers is not in the document]
Comment (mihaimaruseac) [reply pkg/assembler/backends/testing/path.go:148]: Yup! It should not matter much, if this reverse turns out to take too long we can drop it and just draw the arrows reversed in UX

File: pkg/assembler/backends/backends.go (hunk 5 above)
Comment (mihaimaruseac) [reply pkg/assembler/backends/backends.go:77]: It should, but there's a bug in gqlgen that now makes all IDs convert to string scalars. Waiting on a fix there and the next codegen will ask us to resolve

**Pull request merged**